kubelet: fix handle new log file in ReadLogs after log rotation

icefed opened the pull request:
related issues:
https://github.com/kubernetes/kubernetes/issues/59902 https://github.com/kubernetes/kubernetes/issues/28369

<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

/kind bug
<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:

After `Add CRI container log rotation support` https://github.com/kubernetes/kubernetes/pull/59898, we need log API to support `log rotation` and @giuseppe have done in `kubelet: handle recreated log files` https://github.com/kubernetes/kubernetes/commit/341c2c0d1fe13d36c44e41e29c21f46daa0073b8.

But there is a bug in it, new log file may not created when opening it after log file renamed, this will cause `kubectl logs` stuck.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #59902 #28369 

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
none
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
none
```

saschagrunert commented on pkg/kubelet/kuberuntime/logs/logs.go:
```diff
@@ -477,3 +481,18 @@ func waitLogs(ctx context.Context, id string, w *fsnotify.Watcher, runtimeServic
 	}
 }
+
+// retryOpen retries to open the file for 5 times with 1 second interval.
+func retryOpen(path string) (*os.File, error) {
+	var err error
+	for i := 0; i < 5; i++ {
+		var f *os.File
+		f, err = os.Open(path)
+		if err == nil {
+			return f, nil
+		}
+		klog.ErrorS(err, "Failed to open file, retrying", "path", path)
+		time.Sleep(1 * time.Second)
+	}
+	return nil, err
+}
```
I'd say we wait with a backoff here, see:
https://github.com/kubernetes/kubernetes/blob/d220ecb415485980ff02b7a010031bf279e9e067/staging/src/k8s.io/apimachinery/pkg/util/wait/backoff.go#L446-L470

icefed replied:
Done.

giuseppe commented:
> But there is a bug in it, new log file may not created when opening it after log file renamed, this will cause kubectl logs stuck.

how does it happen?  Isn't the file created atomically (using rename)?

icefed commented:
> > But there is a bug in it, new log file may not created when opening it after log file renamed, this will cause kubectl logs stuck.
> 
> how does it happen? Isn't the file created atomically (using rename)?

`containerLogManager` only renames and calls `CRIRumtine.ReopenContainerLog`. The creation timing of a new log file is dependent on the CRI call, and there is a gap of time (less than 1s in my testing) between the file's rename and its creation. Therefore, when `fsnotify.Rename` is received to open the new log file, it may not have been created yet.

bart0sh commented on pkg/kubelet/kuberuntime/logs/logs.go:
```diff
@@ -358,7 +359,11 @@ func ReadLogs(ctx context.Context, path, containerID string, opts *LogOptions, r
 					return err
 				}
 				if recreated {
```
Isn't `rectreated` is set to `true` means that the file exists? If it's set to `true` if file isn't exist then probably `waitLogs` should be fixed.

icefed replied:
If `recreated` is `true` means the log file has beed changed(including being created, renamed and removed), and if log rotation is in progress, CRI will recreate it. 

`recreated` may not be entirely precise here.

bart0sh replied:
waitLogs descriptions states that `recreated` is set to `true` if file is re-created:
```
waitLogs wait for the next log write. It returns two booleans and an error. The first boolean indicates whether a new log is found; the second boolean if the log file was recreated; the error is error happens during waiting new logs.
```
If it's not the case then it probably should be fixed.

My point is that instead of working around the case when file doesn't exist it's better to fix `waitLogs` to return correct info.

BTW, `waitLog` also returns `found`, which means that the new log is found. You can check if `found` is set to `true` instead of trying to open the file 8 times.

icefed replied:
You are correct, we should fix waitLogs instead of retryOpen.

icefed replied:
> @icefed Please, reimplement according to what we've discussed [here](https://github.com/kubernetes/kubernetes/pull/118500#pullrequestreview-1467066749)

@bart0sh `recreated` now returns the correct value. Is there anything else that needs to be modified? Do you mean removing `retrying`?

icefed replied:
If we returning `false, false, nil` on fsnotify.Rename, fsnotify.Create event will not be received.

icefed replied:
More modifications need to be made in ReadLogs to accomplish this.

bart0sh replied:
> If we returning false, false, nil on fsnotify.Rename, fsnotify.Create event will not be received.

That means that fsnotify watcher doesn't watch logs directory, I guess. It probably should. Or it should be reconfigured to watch new file name after renaming.

icefed replied:
Watching logs directory is not ideal as log rotation also involves compression operations.  Reconfigures to new log file is not enough to fix the bug, more modifications need to be made in ReadLogs.

If both `found` and `recreated` are true, `ReadLogs` will return and cannot proceed to the next `waitLogs`.

icefed replied:
I've made some additional modifications and added unit tests.

giuseppe commented:
I had a quick look at the `containerLogManager.rotateLatestLog()` function and I think the issue is in the log file that gets deleted while the log is rotated, so there is a time window where the `log` path is broken.  I'd try solving it in another way, that also takes care of the issue described by the inline comment `// If kubelet gets restarted at this point, we'll lose original log.`

Instead of renaming the file, just create a hard link:

`c.osInterface.Rename(log, rotated)`  becomes `c.osInterface.Link(log, rotated)`

At this point, there are two paths that point to the same file.

Then it is safe to call `ReopenContainerLog` which will create a new empty file and rename it on top of `log`, and there is no need to do the second `Rename` in case of errors but unlink the previously created hard link: `Unlink(rotated)`.

There is no point time where the `log` path doesn't point to an existing file, and even if the Kubelet crashes, the file won't be lost.

What do you think?

giuseppe commented:
> The `ReopenContainerLog` function is implemented by CRI, and there are no additional parameters in the request to customize the file name for reopening, additionally there is no rename method (log rotation is a kubelet behavior). If this were to be changed, it would require modifications to the CRI API and implementation, I don't believe that it would be a better approach.

but `containerLogManager.rotateLatestLog()` is in the Kubelet and that is where the file disappears right?  Why would it need any change in the CRI?  There is no need to customize the target file, it can reuse the same file name as it does now.

icefed commented:
> > The `ReopenContainerLog` function is implemented by CRI, and there are no additional parameters in the request to customize the file name for reopening, additionally there is no rename method (log rotation is a kubelet behavior). If this were to be changed, it would require modifications to the CRI API and implementation, I don't believe that it would be a better approach.
> 
> but `containerLogManager.rotateLatestLog()` is in the Kubelet and that is where the file disappears right? Why would it need any change in the CRI? There is no need to customize the target file, it can reuse the same file name as it does now.

Sorry, it appears that there may be some discrepancy in our understanding of the current implementation of ReopenContainerLog. If we use a hard link for the log and then call ReopenContainerLog, CRI will only open the old file and will not create a new empty file.

https://github.com/containerd/containerd/blob/67349c11411f5bf3b97cf81d0f95d3287816d2cf/pkg/cri/server/container_log_reopen.go#L27C1-L40

icefed commented:
> at least in the case of CRI-O/conmon the new file is not directly opened but created with a tmpfile+rename so it overrides the existing file (https://github.com/containers/conmon/blob/main/src/ctr_logging.c#L613-L621).

`CRI-O/conmon` is not an implementation of the CRI API. We cannot guarantee that all CRI implementations can operate like conmon.

> There shouldn't be a need to retry opening a file on the inotify event since the event happens after. Maybe we are just reacting to the wrong event? e.g. the `rename` event generated by the `rotateLatestLog` should probably be ignored.

The issue still persists even if we ignore the `rename` event.

Thanks, anyway.

bart0sh commented:
/triage accepted
/priority important-soon
/ok-to-test
/assign

@icefed Please, reimplement according to what we've discussed [here](https://github.com/kubernetes/kubernetes/pull/118500#pullrequestreview-1467066749)

bart0sh commented on pkg/kubelet/kuberuntime/logs/logs.go:
```diff
@@ -456,11 +457,19 @@ func waitLogs(ctx context.Context, id string, w *fsnotify.Watcher, runtimeServic
+				// fix issues:
+				// - https://github.com/kubernetes/kubernetes/issues/59902
+				// - https://github.com/kubernetes/kubernetes/issues/28369
+				err := waitFileRecreated(ctx, e.Name)
+				if err != nil {
+					return false, false, err
+				}
+				return true, true, nil
```
```suggestion
				return false, false, nil
```

icefed replied:
The bug still persists even after making these changes.

bart0sh commented on pkg/kubelet/kuberuntime/logs/logs_test.go:
```diff
@@ -454,3 +453,67 @@ func TestReadLogsLimitsWithTimestamps(t *testing.T) {
+func TestReadLogsWithLogRotation(t *testing.T) {
+	logLineFmt := "2023-06-09T05:28:57.806368292Z stdout F %v\n"
+
+	dir, err := os.MkdirTemp("", "readlogs")
+	assert.NoError(t, err)
+	defer os.RemoveAll(dir)
+	tmpfile, err := os.CreateTemp(dir, "log.*.txt")
```
```suggestion
	tmpfile, err := os.CreateTemp(t.TempDir(), "log.*.txt")
```

icefed replied:
It seems that `fsnotify` may lose events if there are insufficient permissions for `t.TempDir()`, at least on my Mac OS. Creating a new directory is ok.

bart0sh replied:
That sounds interesting, considering that `t.TempDir` [creates a directory with `0777` permissions](https://cs.opensource.google/go/go/+/refs/tags/go1.20.5:src/testing/testing.go;drc=38cfb3be9d486833456276777155980d1ec0823e;l=1223). Can you give more details?

icefed replied:
Sorry, I mistakenly interpreted `os.CreateTemp(t.TempDir(), "log.*.txt")` as `os.CreateTemp("", "log.*.txt")`.
It's not fine with `os.CreateTemp("", "log.*.txt")`.

Fixed.

bart0sh commented on pkg/kubelet/kuberuntime/logs/logs_test.go:
```diff
@@ -454,3 +453,67 @@ func TestReadLogsLimitsWithTimestamps(t *testing.T) {
+
+		// call CRI.ReopenContainerLog took some time
+		time.Sleep(500 * time.Millisecond)
+		newTmpfile, err := os.OpenFile(tmpfile.Name(), os.O_CREATE|os.O_APPEND|os.O_WRONLY, 0644)
+		assert.NoError(t, err)
+
+		for i := 5; i < 10; i++ {
+			newTmpfile.WriteString(fmt.Sprintf(logLineFmt, i))
+		}
+		newTmpfile.Close()
```
```
		newTmpfile, err := os.OpenFile(tmpfile.Name(), os.O_CREATE|os.O_APPEND|os.O_WRONLY, 0644)
		assert.NoError(t, err)

		for i := 5; i < 10; i++ {
			newTmpfile.WriteString(fmt.Sprintf(logLineFmt, i))
		}
		newTmpfile.Close()
```
This looks like a good candidate to be wrapped into function and called here and when the file was created first time.

icefed replied:
Done.

bart0sh commented on pkg/kubelet/kuberuntime/logs/logs_test.go:
```diff
@@ -454,3 +453,67 @@ func TestReadLogsLimitsWithTimestamps(t *testing.T) {
+	}
+
+	assert.Equal(t, 10, lineCount, "should have 10 lines")
+}
```
Would it make sense to modify `TestReadLogs` instead of writing new test?

icefed replied:
The handling of this test differs slightly from the case in TestReadLogs.

bart0sh replied:
If it differs slightly it's a good chance to modify it. It looks a bit strange to have to tests testing the same function.

icefed replied:
`TestReadLogs` currently includes some test cases specific to `PodLogOptions`. I suggest separating the tests for different scenarios, since we already have other tests that cover these scenarios independently.

```
TestLogOptions
TestReadLogs
TestParseLog
TestWriteLogs
TestWriteLogsWithBytesLimit
TestReadLogsLimitsWithTimestamps
TestReadLogsWithLogRotation
```

bart0sh replied:
makes sense to me, thanks.

bart0sh commented:
@icefed it's not directly related to the bugfix, but it would be nice to spend some time on readability improvements for the code.  `ReadLogs` function is hard to understand, at least for me. It could be done as a separate PR.

icefed commented:
> @icefed it's not directly related to the bugfix, but it would be nice to spend some time on readability improvements for the code. `ReadLogs` function is hard to understand, at least for me. It could be done as a separate PR.

I'll try next time.

icefed committed:
kubelet: fix handle new log file in ReadLogs after log rotation

related issues:
https://github.com/kubernetes/kubernetes/issues/59902
https://github.com/kubernetes/kubernetes/issues/28369

bart0sh commented:
/retest

icefed commented:
Hi @bart0sh, thanks for your review. 

Are there any other changes that need to be made? or anything else that could impact approval?

bart0sh commented:
@icefed 
> Are there any other changes that need to be made? or anything else that could impact approval?
It looks good to me to pass further. Thanks for the ping.

/lgtm
/assign @mrunalp @derekwaynecarr @dchen1107

tuibeovince commented:
The fixes being implemented in this PR is also very similar to the fixes in https://github.com/kubernetes/kubernetes/pull/115702 

Maybe there is a way to make a unified solution.

tuibeovince commented:
> The fixes being implemented in this PR is also very similar to the fixes in #115702
> 
> Maybe there is a way to make a unified solution.

The aforementioned PR has been merged. Hopefully, that PR's fixes cover the fix for this one too.
